File: src/main/java/org/usfirst/irs1318/gamesim/engine/actors/Field.java
package org.usfirst.irs1318.gamesim.engine.actors;

import org.usfirst.irs1318.gamesim.engine.activities.Direction;
import org.usfirst.irs1318.gamesim.engine.container.Container;

import java.util.*;

public final class Field extends Actor {
    private final int width;
    private final int height;
    private Phase phase;

    /**
     * Containers that can be accessed by location rather than String token.
     */
    private Map<Location, FieldSquare> fieldSquares;

    private Field(int width, int height, Phase phase, Map<Location, FieldSquare> fieldSquares, Location location, Map<String, Container> containers) {
        super(location, containers);
        this.width = width;
        this.height = height;
        this.fieldSquares = fieldSquares;
        this.phase = phase;
    }

    public enum Phase {
        PRE,
        AUTO,
        TELEOP,
        POST
    }

    public Location getDestination(Location start, Direction direction) {
        // TODO
        throw new RuntimeException("Not implemented");
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public Map<Location, FieldSquare> getFieldSquares() {
        return fieldSquares;
    }

    public void setFieldSquares(Map<Location, FieldSquare> fieldSquares) {
        this.fieldSquares = fieldSquares;
    }

    public Phase getPhase() {
        return phase;
    }

    public void setPhase(Phase phase) {
        this.phase = phase;
    }

    public static class Builder extends Actor.Builder {
        private int width;
        private int height;
        private Phase phase = Phase.PRE;
        private Map<Location, FieldSquare> fieldSquares;
        private Location location = new Location(0, 0);
        private Map<String, Container> containers = new HashMap<>();

        public Builder setWidth(int width) {
            this.width = width;
            return this;
        }

        public Builder setHeight(int height) {
            this.height = height;
            return this;
        }

        public Builder setPhase(Phase phase) {
            this.phase = phase;
            return this;
        }

        public Builder setFieldSquares(Map<Location, FieldSquare> fieldSquares) {
            this.fieldSquares = fieldSquares;
            return this;
        }

        public Builder putFieldSquare(Location location, FieldSquare fieldSquare) {
            this.fieldSquares.put(location, fieldSquare);
            return this;
        }

        public Builder setLocation(Location location) {
            this.location = location;
            return this;
        }

        public Builder setContainers(Map<String, Container> containers) {
            this.containers = containers;
            return this;
        }

        public Builder putContainer(String string, Container container) {
            this.containers.put(string, container);
            return this;
        }

        @Override
        public Actor build() {
            return new Field(width, height, phase, fieldSquares, location, containers);
        }
    }
}

File: src/main/scala/org/usfirst/irs1318/gamesim/engine/actors/Field.scala
package org.usfirst.irs1318.gamesim.engine.actors

import org.usfirst.irs1318.gamesim.engine.actors.Field.Contents

case class Field(width: Int,
                 height: Int,
                 phase: Field.Phase,
                 contents: Contents) {
  @inline def mapPhase(fn: Field.Phase => Field.Phase): Field = copy(phase = fn(phase))
  @inline def mapContents(fn: Contents => Contents): Field = copy(contents = fn(contents))
}

object Field {
  type Content = Actor
  type Contents = Map[String, Content]

  sealed abstract class Phase

  object Phase {
    case object Pre extends Phase
    case object Auto extends Phase
    case object Teleop extends Phase
    case object Post extends Phase
  }

  case class Builder(field: Field) {
    def this() = this(Field(0, 0, Phase.Pre, Map.empty))
    @inline def mapField(fn: Field => Field): Builder = copy(field = fn(field))
    def setWidth(width: Int): Builder = mapField(_.copy(width = width))
    def setHeight(height: Int): Builder = mapField(_.copy(height = height))
    def setPhase(phase: Phase): Builder = mapField(_.copy(phase = phase))
    def setContents(contents: Contents): Builder = mapField(_.mapContents(_ => contents))
    def putContent(content: Content): Builder = mapField(_.mapContents(_ + ((content.name, content))))
    def build: Field = field
  }
}
